`kafka/server`: alter `tristate` handling

WillemKauf opened the pull request:
`<= 0` is not a good condition for the `-1` "disabled" case for a `tristate`. `0` is a valid value for a number of `tristate` values, and a `tristate` should only be considered `disabled` for values `< 0`.

For example, `min.cleanable.dirty.ratio` is a `tristate` and `0` is a perfectly acceptable value for it.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

### Bug Fixes

* Fixes a bug in which `tristate` topic properties would be considered disabled if set with the value `0` during a `CreateTopic` or `AlterConfig` request. Now, topic properties set with `0` will reflect that value exactly. This affects the following topic properties: `segment.ms`, `retention.bytes`, `retention.ms`, `retention.local.target.bytes`, `retention.local.target.ms`, `initial.retention.local.target.bytes`, `initial.retention.local.target.ms`, `delete.retention.ms`, `min.cleanable.dirty.ratio`.

WillemKauf committed:
`kafka/server`: alter `tristate` handling for `CreateTopics` commands

`<= 0` is not a good condition for the `-1` "disabled" case for a `tristate`.
`0` is a perfectly valid value for a number of `tristate` values, and a
`tristate` should only be considered `disabled` for values `< 0`.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--fac35fbd-00e9-45d7-bfa2-537af1e2a699-->
<!--TEST RESULTS-->
<!--12cb450b-adea-4085-bd8a-b1035e090280-->
<!--TEST RESULTS-->
<!--5ef62706-9bb4-4b9a-a3f8-4cd5782f69cb-->
<!--TEST RESULTS-->
<!--a66d4ca3-b5a7-4152-933a-c0d92750f0a6-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77224">build#77224</a></summary>


| test_class                 | test_method                           | test_arguments                                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                 |
|:---------------------------|:--------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodePostRestartProbeTest   | post_restart_probe_test               | null                                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/77224#019ae241-f0de-422c-b3bb-eb4707964026 | FLAKY         | 20/21    | upstream reliability is '99.54337899543378'. current run reliability is '95.23809523809523'. drift is 4.30528 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodePostRestartProbeTest&test_method=post_restart_probe_test               |
| NodesDecommissioningTest   | test_decommissioning_rebalancing_node | {"shutdown_decommissioned": true}                                                                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/77224#019ae241-f0de-422c-b3bb-eb4707964026 | FLAKY         | 14/21    | upstream reliability is '100.0'. current run reliability is '66.66666666666666'. drift is 33.33333 and the allowed drift is set to 50. The test should PASS            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_rebalancing_node |
| PartitionReassignmentsTest | test_reassignments_cancel             | null                                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/77224#019ae241-f0dc-4a2a-ab97-ab07ba77e43c | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionReassignmentsTest&test_method=test_reassignments_cancel           |
| TieredStorageTest          | test_tiered_storage                   | {"cloud_storage_type_and_url_style": [1, "virtual_host"], "test_case": {"name": "(TS_Read == True, TS_Timequery == True, SpilloverManifestUploaded == True)"}} | integration | https://buildkite.com/redpanda/redpanda/builds/77224#019ae23c-8232-4340-903c-db2315c22628 | FLAKY         | 20/21    | upstream reliability is '99.20424403183023'. current run reliability is '95.23809523809523'. drift is 3.96615 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TieredStorageTest&test_method=test_tiered_storage                          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77890">build#77890</a></summary>


| test_class          | test_method                                | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                 |
|:--------------------|:-------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CrashLoopChecksTest | test_cloud_storage_client_misconfiguration | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77890#019b2390-b4a3-417b-8ca8-c90df9e81fad | FLAKY         | 1/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.7037, p0=0.1552, reject_threshold=0.0100. adj_baseline=0.9740, p1=0.2317, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CrashLoopChecksTest&test_method=test_cloud_storage_client_misconfiguration |
| CrashLoopChecksTest | test_cloud_storage_client_misconfiguration | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77890#019b2395-fee0-46aa-bdb4-f18c8579ad9d | FAIL          | 0/51     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.8987, p0=0.0048, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CrashLoopChecksTest&test_method=test_cloud_storage_client_misconfiguration |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77920">build#77920</a></summary>


| test_class          | test_method                                | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                 |
|:--------------------|:-------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CrashLoopChecksTest | test_cloud_storage_client_misconfiguration | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77920#019b257b-ce6c-4519-9be8-ea183d6b667f | FAIL          | 0/51     | Test PASSES after retries.[Inconclusive result after max retries](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#Case-F%3A-inconclusive-result-after-max-retries-(run-passes)%5BinlineExtension%5D)(baseline=0.9265, p0=0.0220, reject_threshold=0.0100. adj_baseline=0.9996, p1=1.0000, trust_threshold=0.5000)                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CrashLoopChecksTest&test_method=test_cloud_storage_client_misconfiguration |
| ShardPlacementTest  | test_core_count_change                     | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77920#019b257b-ce6b-42b6-86fb-900274929e86 | FLAKY         | 40/41    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0060, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.0179, p1=0.4850, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShardPlacementTest&test_method=test_core_count_change                      |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77975">build#77975</a></summary>


| test_class            | test_method                      | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                        |
|:----------------------|:---------------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogUpgradeTest   | test_audit_log_upgrade_all_nodes | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/77975#019b2822-3860-4b69-a5f2-961b1548eb59 | FLAKY         | 47/51    | Test PASSES after retries.[Inconclusive result after max retries](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#Case-F%3A-inconclusive-result-after-max-retries-(run-passes)%5BinlineExtension%5D)(baseline=0.0124, p0=0.0242, reject_threshold=0.0100. adj_baseline=0.0367, p1=0.8890, trust_threshold=0.5000)                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogUpgradeTest&test_method=test_audit_log_upgrade_all_nodes  |
| UpgradeBackToBackTest | test_upgrade_with_all_workloads  | {"single_upgrade": false} | integration | https://buildkite.com/redpanda/redpanda/builds/77975#019b2825-67a2-4fbe-b128-71647cee7cf7 | FLAKY         | 30/31    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0092, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.0273, p1=0.4358, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=UpgradeBackToBackTest&test_method=test_upgrade_with_all_workloads |
</details>
<!--/body-->

andrwng commented on src/v/kafka/server/handlers/topics/types.h:
```diff
@@ -190,7 +190,7 @@ get_tristate_value(const config_map_t& config, std::string_view key) {
     }
     // disabled case
-    if (v <= 0) {
+    if (v < 0) {
```
Just calling it out since it wasn't clear to me at first, this commit isn't a functional change because of L189. Still a win for readability though

pgellert replied:
I think this *is* a functional change, right? Or am I missing something non-obvious here?

If the topic property `min.cleanable.dirty.ratio` is set to `0.0`, L189 will not be taken because the property is set (`v != std::nullopt`), and now instead of returning a disabled tristate, we return a tristate with the value `0.0`.

WillemKauf replied:
Correct, this is a functional changes, `!v` is a check for `std::nullopt,` not on the underlying value itself, and the change in return value Gellert noted above is accurate.

dotnwat replied:
@pgellert is it true that this tristate helper operates on external values coming in from kafka clients? if so, i think it places a bit of an extra level of scrutiny here because we can't know if any clients have come to depend on a different functional behavior--that is, our own quirks/bugs are now part of our API. of course this isn't a hard and fast rule, but just something to be aware of.

WillemKauf replied:
The cover letter enumerates the topic properties which have their handling changed as a result of this PR.

It's pretty expansive and mostly affects retention values. It would really suck if someone out there was setting these to 0 and now accidentally wipes their data (though I'm not sure what else a user could expect in doing so).

To be clear, I only became aware of this mishandling because I was trying to set `min.cleanable.dirty.ratio` to `0` for a topic in a ducktape test, which ended up disabling it instead and falling back to the cluster default of `0.2`. 

I agree that it is probably better to maintain status quo instead of fixing a bug that potentially has dramatic side effects for existing users.

dotnwat replied:
> instead of fixing a bug

i think that not being able to set a valid value for an option is also a bug. what we need to do is define the broad classes of concern/scenarios rather than looking at specific cases (specific cases are important too)

pgellert commented on src/v/kafka/server/handlers/topics/types.h:
```diff
@@ -190,7 +190,7 @@ get_tristate_value(const config_map_t& config, std::string_view key) {
     }
     // disabled case
-    if (v <= 0) {
+    if (v < 0) {
```
Is this change correct for the other tristate topic properties as well? For example, `retention.bytes` and `retention.ms` can also be set to 0. Do we know whether treating 0 as valid is correct for all other topic properties that are tristates?

WillemKauf replied:
After some searching it seems valid. If anything, these topic properties should be performing their own validation if `0` were not an acceptable value. What's interesting is at the cluster level, `log_retention_ms` also considers `0` valid, and values `< 0` as "disabling": https://github.com/redpanda-data/redpanda/blob/7f1ac4ffec7a1ae00619d9e721078e7dee8887bc/src/v/config/property.h#L995-L1001

pgellert reviewed (commented):
I think we should add a release note for this, mentioning the topic properties that it impacts.
What are your thoughts on backporting this one? It seems like a bug fix to me, but then I think in the past I think there was a policy that we don't backport topic-property-related changes.

WillemKauf commented:
> I think we should add a release note for this, mentioning the topic properties that it impacts.

Sure, that seems fair.

> What are your thoughts on backporting this one? It seems like a bug fix to me, but then I think in the past I think there was a policy that we don't backport topic-property-related changes.

I don't think its meaningful enough to backport.

WillemKauf commented:
@redpanda-data/core-enterprise does anyone wanna sign off on this?

WillemKauf committed:
`kafka/server`: alter `tristate` handling for `AlterTopic` commands

`<= 0` is not a good condition for the `-1` "disabled" case for a `tristate`.
`0` is a perfectly valid value for a number of `tristate` values, and a
`tristate` should only be considered `disabled` for values `< 0`.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/77890-->
<!--7da0641a-b167-4a1b-8b81-9aba82faf2b7-->
<!--3573787a-74ad-4f74-a536-e28924ab10c1-->
<!--title-->
## Retry command for [Build#77890](https://buildkite.com/redpanda/redpanda/builds/77890)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/crash_loop_checks_test.py::CrashLoopChecksTest.test_cloud_storage_client_misconfiguration
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/77920-->
<!--1ac5026c-1481-4895-8089-ee1f7d6dbe47-->
<!--title-->
## Retry command for [Build#77920](https://buildkite.com/redpanda/redpanda/builds/77920)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/crash_loop_checks_test.py::CrashLoopChecksTest.test_cloud_storage_client_misconfiguration
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/crash_loop_checks_test.py::CrashLoopChecksTest.test_cloud_storage_client_misconfiguration

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/77962-->
<!--fbc060d4-a968-4360-bb46-114b96023067-->
<!--title-->
## Retry command for [Build#77962](https://buildkite.com/redpanda/redpanda/builds/77962)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/crash_loop_checks_test.py::CrashLoopChecksTest.test_cloud_storage_client_misconfiguration
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1
skip-units
tests/rptest/tests/crash_loop_checks_test.py::CrashLoopChecksTest.test_cloud_storage_client_misconfiguration

pgellert commented on src/v/kafka/server/handlers/configs/config_utils.h:
```diff
@@ -608,8 +608,8 @@ void parse_and_set_tristate(
-            property.value = tristate<T>{};
+        if (parsed < 0) {
+            property.value = tristate<T>{disable_tristate};
         } else {
```
This doesn't affect `segment.ms` because of the clamping logic of `log_segment_size_min` and `log_segment_size_max`, right?

WillemKauf replied:
Correct, we clamp in the storage layer in `disk_log_impl::apply_segment_ms()`.

pgellert reviewed (changes requested):
@WillemKauf can you please add a regression test for this? I think we could test this through a fixture test, something like this: https://gist.github.com/pgellert/a5b47e6b5e9d4e7f41ff6c7f82c65a55.

This is a fairly significant behavior change for some topic properties. For example, `delete.retention.ms=0` will now mean delete data immediately rather than keep it indefinitely. Because of that, I’d like to make the behaviour explicit in a test, and then call it out clearly in the release notes so customers are not surprised.

Once this is merged, could you also file a docs ticket to cover the release note update? Thanks.

WillemKauf committed:
`kafka/server`: add `test_tristate_handling_alter_config`

WillemKauf committed:
`kafka/server`: add `test_tristate_handling_create_topic`

WillemKauf commented:
> can you please add a regression test for this?

Added two tests, `test_tristate_handling_alter_config` and `test_tristate_handling_create_topic`.

pgellert reviewed (approved):
lgtm - thanks for the fix!

Please also cc Matt on the docs ticket for awareness.
